File: ruby/leap/leap.rb
class Year

  def self.leap?(yr)
    return false unless divisible_by?(yr, 4)
    if divisible_by?(yr, 100)
      return false unless divisible_by?(yr, 400)
    end
    true
  end

  class << self
    private

    def divisible_by?(dividend, divisor)
      dividend % divisor == 0
    end
  end
end

File: javascript/leap/leap.js
// on every year that is evenly divisible by 4
//   except every year that is evenly divisible by 100
//     unless the year is also evenly divisible by 400

const divisibleBy = (dividend, divisor) => dividend % divisor == 0;

export const isLeap = (year) => {
  return divisibleBy(year, 4) &&
    !(
      divisibleBy(year, 100) &&
      !divisibleBy(year, 400)
    );
};
